### ISSUE TITLE
AIX build breaks for master branch.

### ISSUE THREAD

**Issue by KamathForAIX** [scipy/scipy#19895]
### Describe your issue.

AIX CI build for Scipy has been broken with the following error:.

```
FAILED: scipy/special/lib_amos.a.p/_amos.c.o 
In file included from /usr/include/sys/context.h:33,
                 from /opt/freeware/lib/gcc/powerpc-ibm-aix7.3.0.0/10/include-fixed/sys/signal.h:374,
                 from /opt/freeware/lib/gcc/powerpc-ibm-aix7.3.0.0/10/include-fixed/sys/wait.h:59,
                 from /opt/freeware/lib/gcc/powerpc-ibm-aix7.3.0.0/10/include-fixed/stdlib.h:396,
                 from ../scipy/special/_amos.c:95:
../scipy/special/_amos.c: In function 'amos_seri':
../scipy/special/_amos.c:4203:45: error: expected identifier or '(' before numeric constant
 4203 |     double complex ak1, ck, coef, crsc, cz, hz, rz, s1, s2, w[2];
      |                                             ^~
../
```

This happened because the hz variable is already defined in AIX system as 100. So the compiler replaced hz to 100 and this error came. hz is a macro already defined in AIX. 

In order to fix this, we need either rename the hz variable or undefine the macro if the target is AIX. 

Kindly let me know what you think. 

I tried this. 

```diff
diff -u ../scipy/scipy/special/_amos.c ../scipy/scipy/special/_amos.c_tmp
--- ../scipy/scipy/special/_amos.c      2024-01-17 04:04:02.910092826 -0600
+++ ../scipy/scipy/special/_amos.c_tmp  2024-01-17 04:03:44.944688196 -0600
@@ -95,6 +95,10 @@
 #include <stdlib.h>
 #include "_amos.h"
 
+#if defined(_AIX) && defined(hz)
+#undef hz
+#endif
+
 #ifndef CMPLX
 #define CMPLX(x, y) ((double complex)((double)(x) + I * (double)(y)))
 #endif /* CMPLX */
```

Let me know if you are okay if we can raise a pull request with this fix.

### Reproducing Code Example

```python
git clone https://github.com/scipy/scipy.git
cd scipy
meson setup build  --prefix=$PWD/install_dir -Duse-pythran=false
meson compile -C build -j8 --verbose
```

### Error message

```shell
FAILED: scipy/special/lib_amos.a.p/_amos.c.o 
gcc -pthread -Iscipy/special/lib_amos.a.p -Iscipy/special -I../scipy/special -Iscipy/_lib -I../scipy/_lib -I../scipy/_build_utils/src -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c99 -O2 -g -Wno-unused-but-set-variable -Wno-unused-function -Wno-conversion -Wno-misleading-indentation -maix64 -fPIC -MD -MQ scipy/special/lib_amos.a.p/_amos.c.o -MF scipy/special/lib_amos.a.p/_amos.c.o.d -o scipy/special/lib_amos.a.p/_amos.c.o -c ../scipy/special/_amos.c
In file included from /usr/include/sys/context.h:33,
                 from /opt/freeware/lib/gcc/powerpc-ibm-aix7.3.0.0/10/include-fixed/sys/signal.h:374,
                 from /opt/freeware/lib/gcc/powerpc-ibm-aix7.3.0.0/10/include-fixed/sys/wait.h:59,
                 from /opt/freeware/lib/gcc/powerpc-ibm-aix7.3.0.0/10/include-fixed/stdlib.h:396,
                 from ../scipy/special/_amos.c:95:
../scipy/special/_amos.c: In function 'amos_seri':
../scipy/special/_amos.c:4203:45: error: expected identifier or '(' before numeric constant
 4203 |     double complex ak1, ck, coef, crsc, cz, hz, rz, s1, s2, w[2];
      |                                             ^~
../scipy/special/_amos.c:4223:12: error: lvalue required as left operand of assignment
 4223 |         hz = 0.5*z;
      |            ^
../scipy/special/_amos.c:4269:13: error: 's1' undeclared (first use in this function); did you mean 'ss'?
 4269 |             s1 = 1.0;
      |             ^~
      |             ss
../scipy/special/_amos.c:4269:13: note: each undeclared identifier is reported only once for each function it appears in
../scipy/special/_amos.c:4287:13: error: 's2' undeclared (first use in this function); did you mean 'ss'?
 4287 |             s2 = s1 * coef;
```

### SciPy/NumPy/Python version and system information

```shell
Master branch

https://github.com/scipy/scipy.git
```

**Comment by lucascolley** [comment]
cc @ilayn

**Comment by ilayn** [comment]
Why is AIX defining things globally? It is not a macro. We can't undefine every variable we are trying to use. There is something that AIX is doing apparently.

**Comment by KamathForAIX** [comment]
I understand your concern, @ilayn.

In this case, the file _amos.c includes the stdlib.h part of gcc, which includes wait.h, and wait.h includes signal.h.

signal.h in gcc is including sys/context.h, and this sys/context.h includes the sys/m_params.h defined in /usr/include.

In m_params.h, we have this variable hz defined to represent ticks per second on a clock. This is where the build gets messed up.

So to summarize, in our _amos.c file context, the hz variable will be treated as a constant since it has already been included via stdlib.h. 

From your knowledge, what can we do in this situation? Kindly let me know.

**Comment by ilayn** [comment]
```C
#include <stdio.h>

/* global declaration */
int hz = 100;

int main () {
  /* local declaration */
  int hz = 10;
   printf ("value of hz = %d\n",  hz); /* prints 10 */
   return 0;
}
```
First I am not a power user of C. So take this as a weak explanation. 

What I am trying to emphasize that local and global variable name clash is OK. This code above runs and there is no problem. Because variables can take different values in different scope. 

What you are showing here (I think) is that a preprocessor running over the code and replaces every instance of the variable with its actual value. That is different than variable declaration. That's probably a macro defined somewhere and not a variable. https://en.cppreference.com/w/cpp/preprocessor/replace and macros should not deal with lowercase letters and should be explicitly named.  Hence this one you should ask AIX folks first I'd say.

**Comment by KamathForAIX** [comment]
> So to summarize, in our _amos.c file context, the hz variable will be treated as a constant since it has already been included via stdlib.h.

I should have said hz macro here. Sorry for the same.

**Comment by KamathForAIX** [comment]
AIX being one of the oldest operating systems, there are some definitions with lowercase letters in the system headers, which created problems with other open-source packages as well.

In this case, it is the hz macro defined as 100.

**Comment by ilayn** [comment]
True I understand but there must be intermediate steps on AIX system to undef those things in the tool chain, 

We can't keep track of individual things in the files depending on the build system. It's their prerogative to assess what they are doing and whether they should keep doing things that are not compliant with the best practices.

**Comment by ev-br** [comment]
I think we could accept a patch renaming the hz variable to half_z or something in the amos C sources if that unblocks AIX.

**Comment by KamathForAIX** [comment]
Thank you for the feedback. I will raise a PR soon, renaming hz to half_z. 

Okay @ilayn. I will check with the core team of AIX.

**Comment by rgommers** [comment]
Agreed. We can't test AIX (and NetBSD, and ...) beforehand, but we've always accepted straightforward patches to make things work again.
